### PR TITLE
chore: add UserDescriptor to branches.list

### PR DESCRIPTION
**Pull request by timc1**
While working on [Abstract Home](https://goabstract.atlassian.net/browse/PLATFORM-1601)'s `My Active Branches`, I needed to grab all the branches from a specific user.

This PR adds a `UserDescriptor` as an option to the `branches.list` descriptor, so we can pass `userId` to the query params and get all branches for that user.

### CODE REVIEW

File: src/endpoints/Branches.js [hunk 1]
```diff
@@ -58,17 +59,35 @@ export default class Branches extends Endpoint {
...
+            headers
+          });
+        }
 
-        const response = await this.apiRequest(requestUrl, { headers });
         return wrap(response.data.branches, response);
       },
 
       cli: async () => {
-        if (!descriptor) {
```
Comment (amccloud) [inline src/endpoints/Branches.js:71]: If cli doesn't support fetching branches for a user, we'll want to add a case to warn for DX

Edit: nvm I realize this case is what I suggested, inverted

File: src/endpoints/Branches.js [hunk 2]
```diff
@@ -58,17 +59,35 @@ export default class Branches extends Endpoint {
...
+        } else if (descriptor && descriptor.projectId) {
+          response = await this.apiRequest(
+            `projects/${descriptor.projectId}/branches/?${query}`,
+            { headers }
+          );
+        } else {
+          response = await this.apiRequest(`branches/?${query}`, {
+            headers
+          });
+        }
```
Comment (tommoor) [inline src/endpoints/Branches.js:84]:
```suggestion
        if (descriptor && descriptor.userId) {
          query = querystring.stringify({
            ...queryOptions,
            userId: descriptor.userId
          });
        }

        if (descriptor && descriptor.projectId) {
          response = await this.apiRequest(
            `projects/${descriptor.projectId}/branches/?${query}`,
            { headers }
          );
        } else {
          response = await this.apiRequest(`branches/?${query}`, {
            headers
          });
        }
```

Bonus: This lets you search for a users branches within a project?

File: src/endpoints/Branches.js [hunk 3]
```diff
@@ -58,17 +59,32 @@ export default class Branches extends Endpoint {
...
-          : `branches/?${query}`;
+        const queryOptions = { limit, offset, filter, search };
+        let query = querystring.stringify(queryOptions);
+        let response = null;
+
+        if (descriptor && descriptor.userId) {
+          query = querystring.stringify({
+            ...queryOptions,
+            userId: descriptor.userId
+          });
```
Comment (amccloud) [inline src/endpoints/Branches.js:70]: I get the intention here and it'll work until you're using types

`list({ projectId, userId })` is not currently allowed so `userId` can't be present in the condition below. So `query` is only shared in the else branch but not for the projects branch

If we want to allow filtering branches for a user in a project you'll also want:

`descriptor?: ProjectDescriptor | ProjectMembershipDescriptor | UserDescriptor`

where `type ProjectMembershipDescriptor = {| projectId: string, userId: string |}`

It looks like that type exist but might be [mislabeled and mixed with `OrganizationProjectDescriptor`](https://sdk.goabstract.com/docs/abstract-api/#organizationprojectdescriptor) which doesn't appear to be used

**Comment by timc1** [comment]
@amccloud thanks for the thorough review! I went ahead and updated the type definition, docs, as well as the descriptor mismatch.

File: docs/abstract-api.md [hunk 4]
```diff
@@ -226,6 +226,25 @@ abstract.branches.list({
 });
 ```
 
+List all branches for a user in a project
```
Comment (tommoor) [inline docs/abstract-api.md:229]:
```suggestion
List all active branches for a user in a project
```

File: src/endpoints/Branches.js [hunk 5]
```diff
@@ -58,17 +59,32 @@ export default class Branches extends Endpoint {
...
-        const query = querystring.stringify({ limit, offset, filter, search });
-        const requestUrl = descriptor
-          ? `projects/${descriptor.projectId}/branches/?${query}`
-          : `branches/?${query}`;
+        const queryOptions = { limit, offset, filter, search };
+        let query = querystring.stringify(queryOptions);
+        let response = null;
+
+        if (descriptor && descriptor.userId) {
+          query = querystring.stringify({
```
Comment (tommoor) [inline src/endpoints/Branches.js:67]: This is fine and I'm just nit'ing but you could avoid running stringify twice and neaten this up a little by just adding `userId` to the queryOptions inside of this block and then run  `querystring.stringify` afterwards?

Comment (timc1) [reply src/endpoints/Branches.js]: Ty https://github.com/goabstract/abstract-sdk/pull/293/commits/7c445f02b54fcd018656e7038e41e20a20e27dc8

**Pull request merged**